Navigation: Improve UX on adding links

jeryj opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/19481

## Description
The link control covers up the submenu link + when adding submenu link items. It means that to access the + to add another submenu link, you have to:
- click off of the link to close the link control, 
- then reclick the top navigation item you were working with (because clicking off it closes the submenu list)
- click the +

This PR will be a WIP on testing out different methods. The first one is to move the focus back to wherever it came from after a link is selected on the link control. This means most of the time it ends up on the link text, but if you edit the link from the Link button in the toolbar, it will return there after you make a change.

This is very much a WIP. Currently a kind of playground and discussion for UX iterations on this.

## How has this been tested?
Manually

## Screenshots
![return focus on link change](https://user-images.githubusercontent.com/967608/72473599-ac217d80-37ac-11ea-9109-2621d72664c5.gif)

## Types of changes
<!-- Bug fix (non-breaking change which fixes an issue) -->


## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

shaunandrews commented:
This feels nice and more natural. I wonder, could we try selecting the label text rather than just putting the cursor focus at the end? Especially for URLs, I can imagine wanting to completely replace the label by typing right away.

jeryj commented:
@shaunandrews I like this interaction a lot more as well. I'll try the auto select of the label. I can see how it would be useful to have that, but I could also see how you may not realize it's selected, press a key, and be a bit annoyed it just deleted your label.

jeryj commented:
@shaunandrews The latest is selecting the entire label text (with the cursor/caret at the end of the text). Selecting text hides the caret though, so it doesn't make it totally obvious what's going on. I personally prefer the caret just placed at the far end of the text so you can delete or add from there.

To quickly change between modes (selecting the label vs placing the caret at the end of the text) go to `packages/block-library/src/navigation-link/edit.js` and set line `71` to false:
```
const _experimentalSelectLabelContentOnChange = true; // true selects the text, false places the caret at the end
```

shaunandrews commented:
Very cool. I like highlighting the text when adding a link, it feels really natural to me. However, it doesn't seem to work in Safari.

![highlight in safari](https://user-images.githubusercontent.com/191598/72627655-e27d0b00-391a-11ea-9d6d-e528efe791b0.gif)

Here's how it looks in Chrome:

![highlight in chrome](https://user-images.githubusercontent.com/191598/72627850-2cfe8780-391b-11ea-8aec-3eb96e200c95.gif)

karmatosed commented:
I think this is a definite improvement. Let me remove the feedback label as we seem to have it for design - great iteration.

jeryj commented:
I have this code in a much better spot. Works in Safari now as well. I've tested Firefox, Safari, and Chrome. Note that the + on adding submenu items is gone due to a new bug of some kind unrelated to this commit: https://github.com/WordPress/gutenberg/issues/19788

jeryj committed:
Focus link text when link url changes and LinkControl popover is open

jeryj committed:
Bunk change to rerun travis

obenland commented:
> I like highlighting the text when adding a link

Interesting, I feel the opposite way. To me it feels like I'm supposed to replace it with something (because why else would the text be highlighted?).

How about not setting the cursor at all and just having the navigation-item be focused?

jeryj commented:
> To me it feels like I'm supposed to replace it with something (because why else would the text be highlighted?).
I agree with this take for when you select a page from the list. I like the highlighting when you type in a URL though.

I imagine I could add a check to see if it's a URL vs an item chosen from the suggestions.

jeryj committed:
Fixed tests by removing the click and select part, as we're already moving focus and selecting now.

jeryj committed:
Selects nav link label text if it's url-like. Focuses if not. Automatically adds url-like labels as the label. Put a few too many things together.

jeryj commented:
I've updated a few things:
1. When entering text/url from the add link popover (so, not choosing an option from the suggestions), it will automatically add whatever you entered in the field.
2. If this text looks url-like, the label text will be automatically selected
3. If this text does NOT look url-like, it will place the caret at the end of the text

![focus vs select text](https://user-images.githubusercontent.com/967608/72937592-07181f00-3d2f-11ea-9236-a9c4e1fb4745.gif)

obenland commented:
Nice job, those recent updates feel like they're going in the right direction. 

I can see that you'd want the entire text highlighted when inserting a URL to make it easer add a custom label. Focussing the navigation item as a whole still feels a bit more universally useful, though. I'd worry that folks accidentally delete their labels after inserting a URL

jeryj commented:
>  I'd worry that folks accidentally delete their labels after inserting a URL

100% agree. I'd rather do the focus on suggestion click and select the text on url insertion. Accidentally deleting the label from a suggestion (which is likely the label you wanted) will be really common IMO if we always go with selecting all the text.

I'd like to move forward with the current implementation (focus on suggestion click, select text on manual url entry) if that's OK.

obenland commented:
@apeatling Would you mind taking this for a spin? Inserting a URL and a suggested page and share your thoughts?

jeryj committed:
Fixed tests for navigation block.

Test now selects the label and deletes it only if the url does not equal the label.

apeatling commented:
Having tried this I think it's a big improvement, and we should move ahead with it. Highlighting the text feels natural to me, as I would expect folks would want to change the label. Having the URL you've typed rather than nothing is a smoother experience as a whole.

jeryj committed:
Adds @wordpress/dom to package-lock.json

apeatling commented:
Not sure why the other reviewers were removed above, seems like a Github bug.

jeryj commented:
I rebased and then it wouldn't let me push without a pull. That ended up bringing in _all_ the changes from others as new changes from the other files (111 of them 😳) That may have triggered an automatic review process. I then reverted and force pushed, since that stuff wasn't relevant here.

jeryj commented:
@WordPress/gutenberg-core I believe this merge is ready to go, but I don't want to merge it via the command line unless I absolutely have to :) Do y'all have advice on the best way forwards here?

youknowriad commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -78,6 +80,36 @@ function NavigationLinkEdit( {
+	/**
+	 * Focus the navigation link label text and select it.
+	 */
+	function selectLabelText( ) {
```
I feel we already have a way to do this using selectors/actions without messing with refs, selection... right @ellatrix ?

ellatrix replied:
Yes, if you have access to the value (if HTML you need to convert to rich text value), the block client ID and attribute key.

ellatrix replied:
Why does it need to be selected?

jeryj replied:
It's not about getting the value, but to actually select/highlight the text so that someone can start typing delete the label value. This was discussed here: https://github.com/WordPress/gutenberg/pull/19686#issuecomment-575280635

Can I focus the input on the DOM and highlight/select it without using `ref`s? I'm very new to React and Gutenberg, so any direction is helpful and appreciated :)

youknowriad commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -196,8 +229,11 @@ function NavigationLinkEdit( {
-										return label;
+										// If there's no label, add the URL.
+										return escape( normalizedURL );
 									} )(),
```
Very weird code style, but this was here before anyway.

jeryj replied:
Yeah, just trying to sneak in what I needed in the way that made the most sense :) I'm happy to redo it if you think it'd be cleaner a different way.

youknowriad reviewed (commented):
Are we certain about this UX? It seems a bit weird to me to ignore the post-link insertion options (like openInNewTab)?

jeryj commented:
@youknowriad It seemed to be the best of the current options, as there was no way to continue adding links due to the appenders being covered up by the popover. The only way to close the popover was to click to a top-level or sibling. 

It's definitely not 100% perfect, but we did appear to have some consensus that this UX flow was a significant improvement over the current one in regards to creating navigation items, and specifically submenu items.

apeatling commented:
> Are we certain about this UX? It seems a bit weird to me to ignore the post-link insertion options (like openInNewTab)?

It's possible to get back to these via the sidebar, or hitting the link button in the block toolbar. I do think it makes these options less obvious, but the disruption to the flow of adding more links this popover causes is worse.

aduth commented:
A bit of related commentary at: https://github.com/WordPress/gutenberg/pull/19885#issuecomment-578967785

As far as parallel improvements to the current (previous) implementation:

- I intuitively expect to be able to press <kbd>Escape</kbd> after selecting a link to be able to close the Popover. This currently doesn't work correctly. (See #19931)
- From that point, I might expect to be able to insert a new link simply by pressing <kbd>Enter</kbd>, rather than needing to explicitly click on the "+" button.

jeryj committed:
Fixed code formatting after rebasing

jeryj committed:
Removed test for awaiting for Link Control focus after pressing Enter, as the focus should now be on the navigation link text with the Link Control closed

jeryj committed:
Merge branch 'master' into fix/improve-adding-submenu-links

jeryj committed:
Linting fix

jeryj commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -65,23 +65,21 @@ async function updateActiveNavigationLink( { url, label } ) {
-		// within (focusing on the link preview).
-		await page.waitForSelector(
-			':focus.block-editor-link-control__search-item-title'
-		);
```
@aduth I removed this test that was added in https://github.com/WordPress/gutenberg/pull/19931, as with this UX change the `LinkControl` will be automatically closed and the focus moved to the label text.

aduth commented:
Noting that my comment at https://github.com/WordPress/gutenberg/pull/19462#discussion_r373694597 has quite a bit of overlap with the changes here. In that comment, I discuss the problems associated with a user flow where pressing <kbd>Enter</kbd> upon editing a link would move focus back to the block. In a paragraph block, that would have the result of making it impossible to toggle "Opens in New Tab" without using a cursor (i.e. by keyboard, assistive technologies). I worried that might have been introduced here, but it seems to be fine because the "Link" button in the Navigation Link block's toolbar will always open the link popup.

But, ideally these workflows are consistent for editing links either in a paragraph, or for the Navigation Link block. 

Currently, I foresee the following inconsistencies:

- My leaning for addressing this in the paragraph block is that <kbd>Enter</kbd> should keep the popover open and focused. This would basically be the complete opposite of the changes introduced in this pull request (i.e. it was the previous behavior, if I understand correctly).
- [Conventions in other applications](https://en.wikipedia.org/wiki/Control-K) might lead a user to think that <kbd>Cmd+K</kbd> in a Navigation Link block would put them immediately into the editing mode. Right now, <kbd>Cmd+K</kbd> will shift focus into the Popover, but in its "View" mode. While this can serve as a solution to allow "Open in New Tab" to be accessible by keyboard, it might be contrary to what a user's expectations are.
- A paragraph block's "Link" button will behave as removing the active link. The same button in the Navigation Link block will show the link popover instead. This could be okay, considering that it makes less sense to remove a link from a Navigation Link block than it does to remove a link from a selected text in a paragraph.
